src/ts/Query.ts

```diff
@@ -0,0 +1,305 @@
+import { InvalidParameterError } from '#shared/errors';
+import { BaseScalarArrayType, BaseScalarType } from '#shared/schemaType';
+import type { DbDocument, GenericObject } from '#shared/types';
+import type { ModelConstructor } from './compileModel';
+
+// #region Types
+export interface QueryConstructorOptions {
+	/** Skip the first _n_ results */
+	skip?: number;
+	/** Return only _n_ results */
+	limit?: number;
+	/** Sort criteria */
+	sort?: SortCriteria;
+	/** Return only the indicated properties */
+	projection?: string[];
+}
+
+export interface FilterOperators<TValue> {
+	/** Equal */
+	$eq?: TValue;
+	/** Greater than */
+	$gt?: TValue;
+	/** Greater than or equal to */
+	$gte?: TValue;
+	/** Less than */
+	$lt?: TValue;
+	/** Less than or equal to */
+	$lte?: TValue;
+	/** Not equal */
+	$ne?: TValue;
+	/** String containing */
+	$contains?: TValue;
+	/** String starts with */
+	$startsWith?: TValue;
+	/** String ends with */
+	$endsWith?: TValue;
+	/** In list */
+	$in?: TValue[];
+	/** Not in list */
+	$nin?: TValue[];
+}
+
+export interface RootFilterOperators<TSchema> {
+	/** Used to combine conditions with an and */
+	$and?: Filter<TSchema>[];
+	/** Used to combine conditions with an or */
+	$or?: Filter<TSchema>[];
+}
+
+export type Condition<TValue> = TValue | TValue[] | FilterOperators<TValue>;
+
+export type Filter<TSchema extends GenericObject = GenericObject> = {
+	[key in keyof TSchema]?: Condition<TSchema[key]>;
+} & RootFilterOperators<TSchema>;
+
+export type SortCriteria = [string, -1 | 1][];
+
+export interface QueryExecutionResult {
+	/** Number of documents returned */
+	count: number;
+	/** Unformatted documents returned from database */
+	documents: DbDocument[];
+}
+// #endregion
+
+/** A query object */
+class Query<TSchema extends GenericObject = GenericObject> {
+	/** Model constructor to use with query */
+	private Model: ModelConstructor;
+
+	/** String to use as selection criteria in query */
+	private selection: string | null;
+
+	/** String to use as sort criteria in query */
+	private sort: string | null;
+
+	/** Limit the result set to this number of items */
+	private limit?: number;
+
+	/** Skip this number of items in the result set */
+	private skip?: number;
+
+	/** Specify the projection attribute in result set */
+	private projection: string[];
+
+	public constructor(
```

Comment:
Originally this class allowed for modifications of the parameters on the fly (e.g. limit) via dedicated methods.  We've never used them and the `Query` class is never directly exposed to consumers.  This was really a holdover from trying to model this like mongoose.  I simplified things significantly by removing that logic.